opt: use optgen Let expressions

mgartner opened the pull request:
#### optfmt: fix Let formatting

Previously, `optfmt` could awkwardly format the entire expression on one
line except for the closing `)`. For example:

    (Let ($result $ok):(FoldBinary (OpName) $left $right) $ok
    )

This commit fixes the formatting so that the closing `)` will not be the
only element of the expression placed on a separate line. The example
above is now formatted as:

    (Let
        ($result $ok):(FoldBinary (OpName) $left $right) $ok
    )

Release note: None

#### opt: replace Succeeded and OrderingSucceeded with Let expression

The `Succeeded` and `OrderingSucceeded` custom functions have been
removed. Normalization and exploration rules which used these functions
now use `Let` expressions instead. Custom functions called in these
rules return an additional `ok bool` value which is bound in a `Let`
expression and used to determine if the rule matches an expression.

Release note: None

#### opt: simplify GenerateLocalityOptimizedAntiJoin with Let expression

The commit simplifies `GenerateLocalityOptimizedAntiJoin` by using a
`Let` expression. The `GetLocalityOptimizedAntiJoinLookupExprs` custom
function now returns the local expression, remote expression, and an
`ok` boolean, rather than a single `LocalAndRemoteLookupExprs` struct.
These values are bound to variables in the rule using a `Let`
expression.

Release note: None

mgartner committed:
opt: replace Succeeded and OrderingSucceeded with Let expression

The `Succeeded` and `OrderingSucceeded` custom functions have been
removed. Normalization and exploration rules which used these functions
now use `Let` expressions instead. Custom functions called in these
rules return an additional `ok bool` value which is bound in a `Let`
expression and used to determine if the rule matches an expression.

Release note: None

mgartner committed:
opt: simplify GenerateLocalityOptimizedAntiJoin with Let expression

The commit simplifies `GenerateLocalityOptimizedAntiJoin` by using a
`Let` expression. The `GetLocalityOptimizedAntiJoinLookupExprs` custom
function now returns the local expression, remote expression, and an
`ok` boolean, rather than a single `LocalAndRemoteLookupExprs` struct.
These values are bound to variables in the rule using a `Let`
expression.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63647)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63647)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/norm/rules/bool.opt, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63647#-MYH3mtlCEG6-DJAErgU:-MYH3mtlCEG6-DJAErgV:b-34ycjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b866128e60fccfb051fc16099a7b1e6b60abbb/pkg/sql/opt/norm/rules/bool.opt#L159)):*
> ```
>     $left:^(Or)
>     $right:^(Or) &
>         (Let ($conjunct $ok):
> ```

Sorry I didn't think of this in your previous PR, but the alignment here is confusing, it suggests that everything that follows is logically part of the `:` expression. Is it possible to make optfmt prefer putting the variables on a separate line rather than separating at the colon?

```
(Let
  ($conjunct $ok):(FindRedundantConjunct $left $right)
  $ok
)
```
or for longer cases:
```
(Let
  ($conjunct $ok):(FindRedundantConjunct
    $left
    $right
  )
  $ok
)
```


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63647)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/norm/rules/bool.opt, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63647#-MYH3mtlCEG6-DJAErgU:-MYH5RCJDuOyqqY07oKR:be9n74w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b866128e60fccfb051fc16099a7b1e6b60abbb/pkg/sql/opt/norm/rules/bool.opt#L159)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Sorry I didn't think of this in your previous PR, but the alignment here is confusing, it suggests that everything that follows is logically part of the `:` expression. Is it possible to make optfmt prefer putting the variables on a separate line rather than separating at the colon?

```
(Let
  ($conjunct $ok):(FindRedundantConjunct $left $right)
  $ok
)
```
or for longer cases:
```
(Let
  ($conjunct $ok):(FindRedundantConjunct
    $left
    $right
  )
  $ok
)
```

</blockquote></details>

Most things are possible with the `pretty` package after enough fist shaking and cursing :wink: 

I'll fix in a separate PR. What's your preference when the bind variables are too long to fit on a line with the function?

Something like this:

```
(Let
    ($longVariable1 $veryLongVariable2 $otherLongVariable3 $ok):
        (DoSomething $left $right)
    $ok
)
```

Of course the variable line itself may exceed the max line length, which creates more formatting woes. I'll see what I can do.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63647)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/norm/rules/bool.opt, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63647#-MYH3mtlCEG6-DJAErgU:-MYH76pjBGDEa-rHEdAQ:b-eqe7pl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b866128e60fccfb051fc16099a7b1e6b60abbb/pkg/sql/opt/norm/rules/bool.opt#L159)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Most things are possible with the `pretty` package after enough fist shaking and cursing :wink: 

I'll fix in a separate PR. What's your preference when the bind variables are too long to fit on a line with the function?

Something like this:

```
(Let
    ($longVariable1 $veryLongVariable2 $otherLongVariable3 $ok):
        (DoSomething $left $right)
    $ok
)
```

Of course the variable line itself may exceed the max line length, which creates more formatting woes. I'll see what I can do.

</blockquote></details>

That looks fine, I don't think having a ton of long variable names will happen in practice anyway.

If the variable line itself exceeds the limit, I'd do it like this:
```
(Let
    (
        $longVariable1
        $longVariable2
        ...
    ):(DoSomething
```


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63647)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/norm/rules/bool.opt, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63647#-MYH3mtlCEG6-DJAErgU:-MYHRM5cBcnldesqyhKX:bbp0f08) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b866128e60fccfb051fc16099a7b1e6b60abbb/pkg/sql/opt/norm/rules/bool.opt#L159)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

That looks fine, I don't think having a ton of long variable names will happen in practice anyway.

If the variable line itself exceeds the limit, I'd do it like this:
```
(Let
    (
        $longVariable1
        $longVariable2
        ...
    ):(DoSomething
```
</blockquote></details>

#63696


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for fixing all of these! This is much better!

> Reviewed 13 of 13 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63647#-:-MYLG6e_DVD4HfHxJWnS:b-j6gf6q)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/opt/optgen/cmd/optfmt/main.go:
```diff
@@ -402,16 +402,16 @@ func (p *pp) docOnlyExpr(e lang.Expr) pretty.Doc {
+			pretty.NestT(binding),
 			pretty.Line,
-			p.docExpr(e.Result),
+			pretty.NestT(p.docExpr(e.Result)),
```
There may be a bug in `pretty` that prevents this `pretty.NestT` from indenting correctly. Trying to figure out a work-around.

mgartner committed:
optfmt: fix Let formatting

Previously, `optfmt` could awkwardly format the entire expression on one
line except for the closing `)`. For example:

    (Let ($result $ok):(FoldBinary (OpName) $left $right) $ok
    )

This commit fixes the formatting so that the closing `)` will not be the
only element of the expression placed on a separate line. The example
above is now formatted as:

    (Let
        ($result $ok):(FoldBinary (OpName) $left $right) $ok
    )

Release note: None

mgartner commented:
@RaduBerinde @rytaft PTAL. I finally got around to fixing `optfmt` for `Let` expressions.

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63647#-:-MZtTgLV64FHuZTATLKe:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+
